src/script/lua_api/l_env.cpp

```diff
@@ -841,32 +842,62 @@ int ModApiEnvMod::l_emerge_area(lua_State *L)
 	v3s16 bpmax = getNodeBlockPos(read_v3s16(L, 2));
 	sortBoxVerticies(bpmin, bpmax);
 
-	size_t num_blocks = VoxelArea(bpmin, bpmax).getVolume();
-	assert(num_blocks != 0);
-
-	if (lua_isfunction(L, 3)) {
+	int param = 3;
+	if (lua_isfunction(L, param)) {
 		callback = LuaEmergeAreaCallback;
 
-		lua_pushvalue(L, 3);
+		lua_pushvalue(L, param);
 		int callback_ref = luaL_ref(L, LUA_REGISTRYINDEX);
 
-		lua_pushvalue(L, 4);
+		lua_pushvalue(L, param + 1);
 		int args_ref = luaL_ref(L, LUA_REGISTRYINDEX);
 
 		state = new ScriptCallbackState;
 		state->script       = getServer(L)->getScriptIface();
 		state->callback_ref = callback_ref;
 		state->args_ref     = args_ref;
-		state->refcount     = num_blocks;
+		state->refcount     = -1;
 		state->origin       = getScriptApiBase(L)->getOrigin();
+		param += 2;
 	}
 
-	for (s16 z = bpmin.Z; z <= bpmax.Z; z++)
-	for (s16 y = bpmin.Y; y <= bpmax.Y; y++)
-	for (s16 x = bpmin.X; x <= bpmax.X; x++) {
+	std::string emerge_unit_name = "block";
+	int emerge_unit = 1;
+	if (!lua_isnoneornil(L, param)) {
+		emerge_unit_name = "--INVALID--";
+		if (lua_isstring(L, param))
+			emerge_unit_name = lua_tostring(L, param);
+		if (emerge_unit_name != "chunk" && emerge_unit_name != "block") {
+			luaL_error(L, "emerge_area(): invalid unit. Expected 'block' or 'chunk'");
+			return 1;
+		}
+	}
+	if (emerge_unit_name == "chunk") {
+		emerge_unit = emerge->mgparams->chunksize;
+		bpmin = emerge->getContainingChunk(bpmin);
+		bpmax = emerge->getContainingChunk(bpmax);
+		bpmax.X += emerge_unit - 1;
+		bpmax.Y += emerge_unit - 1;
+		bpmax.Z += emerge_unit - 1;
```

Comment:
Vector3D supports the operation `bpmax += emerge_unit - 1` when `emerge_unit` is `s16`. (`mgparams->chunksize` has the same type)
http://irrlicht.sourceforge.net/docu/vector3d_8h_source.html